src/dataflow/render/mod.rs

```diff
@@ -484,23 +494,22 @@ pub(crate) fn build_dataflow<A: Allocate>(
                             err_collection = err_collection.concat(&errors);
                         }
 
-                        // Apply `as_of` to each timestamp
-                        collection = collection.delay({
-                            let as_of_frontier = as_of_frontier.clone();
-                            move |time| {
-                                let mut time = time.clone();
-                                time.advance_by(as_of_frontier.borrow());
-                                time
-                            }
-                        });
-                        err_collection = err_collection.delay({
-                            let as_of_frontier = as_of_frontier.clone();
-                            move |time| {
-                                let mut time = time.clone();
-                                time.advance_by(as_of_frontier.borrow());
-                                time
-                            }
-                        });
+                        // Apply `as_of` to each timestamp.
+                        let as_of_frontier1 = as_of_frontier.clone();
+                        collection = collection
+                            .inner
+                            .map_in_place(move |(_, time, _)| {
+                                time.advance_by(as_of_frontier1.borrow())
+                            })
+                            .as_collection();
```

Comment:
Is this change just for making non-upsert code faster?

Reply:
No, I think it may have had the same defect: if we have lots of updates with the same time and those updates should compact down, the delay operator would prevent them from compacting in place.